elk/run.py

```diff
@@ -88,28 +90,36 @@ def prepare_data(
 
         return x0, x1, val_x0, val_x1, train_labels, val_labels
 
-    def apply_to_layers(self, func):
-        """Apply a function to each layer of the dataset in parallel."""
-
-        devices = select_usable_devices(self.cfg.num_gpus)
-        num_devices = len(devices)
+    def apply_to_layers(
+        self,
+        func: Callable[[int], Log],
+        num_devices: int,
+        to_csv_line: Callable[[Log], list[str]],
+        csv_columns: list[str],
+    ):
+        """Apply a function to each layer of the dataset in parallel
+        and writes the results to a CSV file.
+
+        Args:
+            func: The function to apply to each layer.
+                The int is the index of the layer.
+            num_devices: The number of devices to use.
+            to_csv_line: A function that converts a Log to a list of strings.
+                This has to be injected in because the Run class does not know
+                the extra options e.g. skip_baseline to apply to function.
+            csv_columns: The columns of the CSV file."""
         self.out_dir = assert_type(Path, self.out_dir)
```

Comment:
i made func have a signature like
```
Callable[[int], Log]
```

previously it was something like

```
Callable[[int, int, list[int]], Log] # layer, device, list of devices
```

which could be pretty confusing

Reply:
previously func wasn't annotated so pyright wouldn't catch keyword mispelling / mistakes in `partial`. 

we could also make it a paramspec instead so you indiate that is a function that has keywords world_size, devices, but i didn't want to complicate things